### DISCUSSION TITLE
How to collect jvm metric: jvm_thread_state on jdk8

### DISCUSSION THREAD

**Discussion by zhangjiabin1010** [open-telemetry/opentelemetry-java-instrumentation#12614] [Q&A] [answered]
Currently jvm.thread.count metric attribute only have jvm.thread.daemon, I need jvm_thread_state 
I see someone has raised this issue,
https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11980
But I don't know how to implement this requirement, can you give some suggestions?

**Comment by laurit** [comment]
`isDaemon` method in `java.lang.management.ThreadInfo` was introduced in jdk9. You could attempt calling `Thread.getAllThreads` via reflection or you could try using `Thread.currentThread().getGroup()` and call `getParent()` until you have determined the root group and use the `enumerate` method to get all the threads. I'd guess that from the `Thread` objects you can extract the necessary information to implement the `jvm.thread.count` metric.

**Reply by zhangjiabin1010** [reply]
> `isDaemon``java.lang.management.ThreadInfo`jdk9 中引入了方法。您可以尝试`Thread.getAllThreads`通过反射进行调用，也可以尝试使用`Thread.currentThread().getGroup()`并调用，`getParent()`直到确定根组并使用该`enumerate`方法获取所有线程。我猜`Thread`您可以从对象中提取实现`jvm.thread.count`度量所需的信息。

OK, these are the methods to get the thread status. I want to integrate this functionality into javaagent-instrumentation.jar and write it to the attribute of jvm.thread.count metric. Where should I start? Can this functionality be achieved through extensions?

**Reply by steverao** [reply, marked as answer]
> I want to integrate this functionality into javaagent-instrumentation.jar and write it to the attribute of jvm.thread.count metric. Where should I start? Can this functionality be achieved through extensions?

You can refer to [some similar modules](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/runtime-telemetry/runtime-telemetry-java8/library/src/main/java/io/opentelemetry/instrumentation/runtimemetrics/java8) to achieve it in OTel Java Agent.
